### PR TITLE
Realm transactions

### PR DESCRIPTION
**Pull request by janherich**
[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

This PR implements transaction compacting for all realm writes/updates/deletes. Previously, we performed each distinct realm operation inside `.write` block, even when we performed multiple such in span of one handler (like adding message, where message needs to be written, chat updated, etc.).
Not only was that not very effective (`.write` blocks denoting transactions are expensive):
screen shot 2018-05-02 at 09 55 05
 
It was also logically wrong, because each handler app-db change is atomic operation (it either succeeds or fails), but that was not true for realm operations, where it could happen that some of the realm objects were written correctly, while some of them not, leading to inconsistent (persisted) state.

This PR address the point by adding `:data-source/tx` and `:data-source/base-tx` effects (for account/base realm respectively) + reworking the `data-source.ENTITY` namespaces to expose public functions modelling transaction operations.
`merge-fx` macro was updated to take special care of those effects, collecting all the transaction operations and issuing them as one transaction (one realm `.write` block) at the end.
I also removed all the `data-source.realm.ENTITY` namespace, as that was just unnecessary burden, which resulted in quite a big code size reduction (almost 300LOC less).

### Testing notes (optional):
Complete regression testing with focus on persistence, eq if things are behaving correctly after app restart/s

status: ready

### CODE REVIEW

**Comment by rasom** [comment]
What is the impact on performance?

File: src/status_im/chat/events.cljs [hunk 1]
```diff
@@ -194,10 +198,10 @@
                         {}
                         all-stored-chats)]
       (handlers-macro/merge-fx cofx
-                         {:db (assoc db :chats chats)}
-                         (init-console-chat)
-                         (group.events/add-default-groups)
-                         (add-default-contacts)))))
+                               {:db (assoc db :chats chats)}
```
Comment (rasom) [inline src/status_im/chat/events.cljs]: @janherich let's just skip re-indentation, we can fix this in a separate PR. By fixing `merge-fx` you don't fix all indentations related problems in project but this adds a lot of noise (PR is big even without this).

Comment (rasom) [reply src/status_im/chat/events.cljs]: Optionally we can put re-indentation into separate commit.

File: src/status_im/chat/events.cljs [hunk 2]
```diff
@@ -290,12 +295,12 @@
 (defn start-chat
   "Start a chat, making sure it exists"
   [chat-id opts {:keys [db] :as cofx}]
-  ; don't allow to open chat with yourself
+                                        ; don't allow to open chat with yourself
```
Comment (rasom) [inline src/status_im/chat/events.cljs]: something is wrong with indentation here

File: src/status_im/chat/models/message.cljs [hunk 3]
```diff
@@ -285,13 +284,13 @@
 (defn send-command
   [{{:keys [current-public-key chats] :as db} :db :keys [now] :as cofx} params]
   (let [{{:keys [handler-data to-message command] :as content} :command chat-id :chat-id} params
-        ; We send commands to deleted chats as well, i.e. signed later transactions
+                                        ; We send commands to deleted chats as well, i.e. signed later transactions
```
Comment (rasom) [inline src/status_im/chat/models/message.cljs]: again comment's indentation

**Review by rasom — CHANGES_REQUESTED** [review]
Please separate changes related to new functionality from indentation changes. It takes too long to find the code which actually requires review.

**Comment by rasom** [comment]
```
FAIL in (upsert-chat-test) (cljs/test.cljs:476:9)
upserting a non existing chat it adds the fx to store a chat
expected: store-chat-fx
  actual: nil

FAIL in (upsert-chat-test) (cljs/test.cljs:476:9)
upserting an existing chat it adds the fx to store a chat
expected: store-chat-fx
  actual: nil

FAIL in (add-group-chat) (cljs/test.cljs:476:9)
it saves the chat in the database
expected: store-fx
  actual: nil

FAIL in (add-public-chat) (cljs/test.cljs:476:9)
it saves the chat in the database
expected: store-fx
  actual: nil
```
tests are failing

File: src/status_im/chat/events.cljs (hunk 2 above)
Comment (janherich) [reply src/status_im/chat/events.cljs]: Fixed

File: src/status_im/chat/events.cljs (hunk 1 above)
Comment (janherich) [reply src/status_im/chat/events.cljs]: I'm little bit uneasy about that, as the indentation is now correct and was wrong previously (when `handlers` -> `handlers-macro` change was made) + github highlights indentation only changes anyway, but if it's necessary, we can do that in such two step approach, sounds like a lot of unnecessary work IMO.

Comment (rasom) [reply src/status_im/chat/events.cljs]: > sounds like a lot of unnecessary work IMO

yep i agree, i wouldn't touch them in the first place because of this. You can't separate visually changes which are only indentations and which change functionality, this means that you literally need to read all `+730 −1,009` in order to review, idk, 100-200 loc. That's a lot of unnecessary work IMO.

**Comment by janherich** [comment]
@rasom Done, old (incorrect) indentation restored, it's now ~200LOC less to review

File: src/status_im/chat/events.cljs (hunk 2 above)
Comment (rasom) [reply src/status_im/chat/events.cljs]: is it?

File: src/status_im/chat/events.cljs [hunk 4]
```diff
@@ -340,8 +256,8 @@
   [re-frame/trim-v]
   (fn [cofx [chat-id]]
     (handlers-macro/merge-fx cofx
-                       (models/remove-chat chat-id)
-                       (navigation/replace-view :home))))
+                             (models/remove-chat chat-id)
+                             (navigation/replace-view :home))))
```
Comment (rasom) [inline src/status_im/chat/events.cljs]: indentation changes are still here

File: src/status_im/chat/models/message.cljs [hunk 5]
```diff
@@ -94,7 +93,7 @@
   [{:keys [chat-id message-id] :as message} {:keys [now] :as cofx}]
   (handlers-macro/merge-fx cofx
                      (chat-model/upsert-chat {:chat-id chat-id
-                                              ; We activate a chat again on new messages
+                                              ;; We activate a chat again on new messages
```
Comment (rasom) [inline src/status_im/chat/models/message.cljs]: why we need this change in PR this particular PR?

**Comment by rasom** [comment]
sorry @janherich, but you added a bunch of other unnecessary changes with `merge-fx`

File: src/status_im/chat/models/message.cljs (hunk 5 above)
Comment (janherich) [reply src/status_im/chat/models/message.cljs]: Because `;` is only used for after line comments, it's explicitly stated in clojure style guide as well.
My mac is setup to auto indent according to those rules and when, so either I don't use auto-indent (sucks very hard), or I introduce indentation changes when I use it.
Same reason why leaving bad indentation with `handlers-macro` (it's inconsistent even among one file, like this one) is bad idea I think and it should be done right (indentation) on the first time, or in the very next PR touching the file (this one). 
When I leave it as it is, it's just the next unfortunate guy touching the files who will auto-indent (according to clojure style guide we are supposed to adhere to) and introduce "unnecessary changes".

**Comment by janherich** [comment]
@rasom Please review again, all the incorrect indentation & comments notation was introduced back and I didn't touch the auto-indent, so hopefully, it will be couple of lines less to review.

**Comment by janherich** [comment]
@rasom Added screenshot illustrating perf benefits on iOS simulator. I still think that the bigger gain is consistency and correctness, but performance is important as well.

File: src/status_im/chat/models/message.cljs (hunk 5 above)
Comment (rasom) [reply src/status_im/chat/models/message.cljs]: Yes that's clear, but you always can create another PR which objective will be style fixes. 
Regarding `handlers-macro` thing: don't worry, i will be that guy and i will file a separate PR with fixed indentations.

File: src/status_im/utils/handlers_macro.cljs [hunk 6]
```diff
@@ -1,16 +1,23 @@
...
+(def ^:private tx-keys #{:data-store/tx :data-store/base-tx})
+
 (defn safe-merge [fx new-fx]
   (if (:merging-fx-with-common-keys fx)
     fx
-    (let [common-keys (clojure.set/intersection (into #{} (keys fx))
-                                                (into #{} (keys new-fx)))]
-      (if (empty? (disj common-keys :db))
-        (merge fx new-fx)
+    (let [common-keys (set/intersection (into #{} (keys fx))
```
Comment (rasom) [inline src/status_im/utils/handlers_macro.cljs:15]: Should we add some comments to this changes? Or probably to `merge-fx`'s doc string as it's behaviour is changed.

File: src/status_im/ui/screens/group/chat_settings/events.cljs [hunk 7]
```diff
@@ -6,15 +6,17 @@
...
+            [status-im.data-store.messages :as messages-store]))
 
 
 ;;;; Handlers
 
 (handlers/register-handler-fx
   :show-group-chat-profile
   [re-frame/trim-v]
-  (fn [{:keys [db] :as cofx} [chat-id]] 
+  (fn [{:keys [db] :as cofx} [chat-id]]
```
Comment (rasom) [inline src/status_im/ui/screens/group/chat_settings/events.cljs:19]: unnecessary change

File: src/status_im/ui/screens/contacts/events.cljs [hunk 8]
```diff
@@ -13,17 +13,18 @@
             [status-im.ui.screens.group.events :as group.events]
             [status-im.chat.console :as console-chat]
             [status-im.chat.events :as chat.events]
-            [status-im.chat.models :as chat.models] 
+            [status-im.chat.models :as chat.models]
```
Comment (rasom) [inline src/status_im/ui/screens/contacts/events.cljs:16]: unnecessary change

File: src/status_im/ui/screens/contacts/events.cljs [hunk 9]
```diff
@@ -4,7 +4,7 @@
             [re-frame.core :as re-frame]
             [status-im.utils.handlers :as handlers]
             [status-im.utils.handlers-macro :as handlers-macro]
-            [status-im.utils.contacts :as utils.contacts]
```
Comment (rasom) [inline src/status_im/ui/screens/contacts/events.cljs:7]: unnecessary change

File: src/status_im/utils/handlers_macro.cljs (hunk 6 above)
Comment (janherich) [reply src/status_im/utils/handlers_macro.cljs:15]: Correct, docstring updated for `merge-fx` macro

File: src/status_im/chat/models/message.cljs (hunk 5 above)
Comment (janherich) [reply src/status_im/chat/models/message.cljs]: Ok, old comment style re-introduced

File: src/status_im/chat/events.cljs [hunk 10]
```diff
@@ -20,9 +20,12 @@
             [status-im.transport.message.v1.protocol :as protocol]
             [status-im.transport.message.v1.public-chat :as public-chat]
             [status-im.transport.message.v1.group-chat :as group-chat]
+            [status-im.data-store.chats :as chats-store]
+            [status-im.data-store.messages :as messages-store]
+            [status-im.data-store.contacts :as contacts-store]
             status-im.chat.events.commands
             status-im.chat.events.requests
-            status-im.chat.events.send-message
```
Comment (rasom) [inline src/status_im/chat/events.cljs:25]: unnecessary change

File: src/status_im/chat/events.cljs [hunk 11]
```diff
@@ -108,7 +112,7 @@
           message (get-in db [:chats chat-id :messages message-id])]
       (models.message/update-message-status message status cofx))))
 
-;; Change status of messages which are still in "sending" status to "not-sent"
+;; Change status of messages which are still in "sending" status to "not-sent
```
Comment (rasom) [inline src/status_im/chat/events.cljs:115]: unnecessary change

File: src/status_im/chat/models.cljs [hunk 12]
```diff
@@ -29,12 +31,12 @@
   [{:keys [chat-id] :as chat-props} {:keys [db] :as cofx}]
   (let [chat (merge
                (or (get (:chats db) chat-id)
-                 (create-new-chat chat-id cofx))
+                   (create-new-chat chat-id cofx))
```
Comment (rasom) [inline src/status_im/chat/models.cljs:34]: unnecessary change

File: src/status_im/chat/models/message.cljs [hunk 13]
```diff
@@ -228,12 +226,12 @@
                         (update :message-type keyword)
                         protocol/map->Message)]
     (handlers-macro/merge-fx cofx
-                             (send chat-id message-id send-record)
-                             (update-message-status message :sending))))
+                       (send chat-id message-id send-record)
```
Comment (rasom) [inline src/status_im/chat/models/message.cljs:229]: unnecessary change

File: src/status_im/chat/models/message.cljs [hunk 14]
```diff
@@ -285,7 +283,7 @@
 (defn send-command
   [{{:keys [current-public-key chats] :as db} :db :keys [now] :as cofx} params]
   (let [{{:keys [handler-data to-message command] :as content} :command chat-id :chat-id} params
-        ; We send commands to deleted chats as well, i.e. signed later transactions
+        ;; We send commands to deleted chats as well, i.e. signed later transactions
```
Comment (rasom) [inline src/status_im/chat/models/message.cljs:286]: unnecessary change

File: src/status_im/data_store/accounts.cljs [hunk 15]
```diff
@@ -1,24 +1,31 @@
...
-            [re-frame.core :as re-frame]
-            [status-im.data-store.realm.core :as core]
-            [status-im.data-store.realm.accounts :as data-store]))
+  (:require [re-frame.core :as re-frame]
+            [status-im.data-store.realm.core :as core]))
 
 ;; TODO janherich: define as cofx once debug handlers are refactored
 (defn get-by-address [address]
-  (data-store/get-by-address address))
+  (-> core/base-realm
```
Comment (rasom) [inline src/status_im/data_store/accounts.cljs]: How is this change related to PR's objective?

File: src/status_im/data_store/accounts.cljs [hunk 16]
```diff
@@ -1,24 +1,31 @@
...
-
-(re-frame/reg-fx
-  :data-store/save-account
-  (fn [{:keys [after-update-event] :as account}]
-    (let [account-to-save (dissoc account :after-update-event)]
-      (async/go (async/>! core/realm-queue #(if after-update-event
-                                              (do (data-store/save account-to-save true)
-                                                  (re-frame/dispatch after-update-event))
-                                              (data-store/save account-to-save true)))))))
+    (assoc coeffects :all-accounts (-> core/base-realm
```
Comment (rasom) [inline src/status_im/data_store/accounts.cljs]: How is this change related to PR's objective?

File: src/status_im/data_store/browser.cljs [hunk 17]
```diff
@@ -1,21 +1,24 @@
...
-            [status-im.data-store.realm.browser :as data-store])
-  (:refer-clojure :exclude [exists?]))
+  (:require [re-frame.core :as re-frame]
+            [status-im.data-store.realm.core :as core]))
 
 (re-frame/reg-cofx
   :data-store/all-browsers
   (fn [cofx _]
-    (assoc cofx :all-stored-browsers (data-store/get-all))))
+    (assoc cofx :all-stored-browsers (-> @core/account-realm
```
Comment (rasom) [inline src/status_im/data_store/browser.cljs:8]: How is this change related to PR's objective?

File: src/status_im/data_store/contacts.cljs [hunk 18]
```diff
@@ -1,42 +1,32 @@
...
 
-(re-frame/reg-fx
-  :data-store/delete-contact
-  (fn [contact]
-    (async/go (async/>! core/realm-queue #(data-store/delete contact)))))
+(defn delete-contact-tx
+  "Returns tx function for deleting contact"
+  [whisper-identity]
+  (fn [realm]
+    (core/delete realm (core/single (core/get-by-field realm :contact :whisper-identity whisper-identity)))))
```
Comment (rasom) [inline src/status_im/data_store/contacts.cljs:32]: 80 chars limit

File: src/status_im/data_store/core.cljs [hunk 19]
```diff
@@ -1,19 +1,36 @@
...
-  (data-source/change-account address new-account? handler))
+  (core/change-account address new-account? handler))
+
+(re-frame/reg-fx
+  :data-store/base-tx
+  (fn [transactions]
+    (async/go (async/>! core/realm-queue (fn []
+                                           (core/write core/base-realm
+                                                       #(doseq [transaction transactions]
+                                                          (transaction core/base-realm))))))))
```
Comment (rasom) [inline src/status_im/data_store/core.cljs]: 80 chars limit

File: src/status_im/data_store/core.cljs [hunk 20]
```diff
@@ -1,19 +1,36 @@
...
+
+(re-frame/reg-fx
+  :data-store/tx
+  (fn [transactions]
+    (async/go (async/>! core/realm-queue (fn []
+                                           (let [realm @core/account-realm]
+                                             (core/write realm
+                                                         #(doseq [transaction transactions]
+                                                            (transaction realm)))))))))
```
Comment (rasom) [inline src/status_im/data_store/core.cljs:39]: 80 chars limit

File: src/status_im/data_store/local_storage.cljs [hunk 21]
```diff
@@ -1,15 +1,17 @@
...
-            [status-im.data-store.realm.local-storage :as data-store]))
+  (:require [re-frame.core :as re-frame]
+            [status-im.data-store.realm.core :as core]))
 
 (re-frame/reg-cofx
   :data-store/get-local-storage-data
   (fn [cofx _]
-    (assoc cofx :get-local-storage-data (comp :data data-store/get-by-chat-id))))
+    (assoc cofx :get-local-storage-data #(-> @core/account-realm
+                                             (core/get-by-field :local-storage :chat-id %)
```
Comment (rasom) [inline src/status_im/data_store/local_storage.cljs:9]: 80 chars limit

File: src/status_im/data_store/messages.cljs [hunk 22]
```diff
@@ -49,7 +49,16 @@
 (re-frame/reg-cofx
   :data-store/message-ids
   (fn [cofx _]
-    (assoc cofx :stored-message-ids (data-store/get-stored-message-ids))))
+    (assoc cofx :stored-message-ids (let [chat-id->message-id (volatile! {})]
```
Comment (rasom) [inline src/status_im/data_store/messages.cljs:52]: Why these changes are necessary for PR?

File: src/status_im/data_store/messages.cljs [hunk 23]
```diff
@@ -59,7 +68,11 @@
            (into {}
                  (map (fn [[chat-id user-statuses]]
                         [chat-id (into #{} (map :message-id) user-statuses)]))
-                 (group-by :chat-id (data-store/get-unviewed (:current-public-key db)))))))
+                 (group-by :chat-id (-> @core/account-realm
```
Comment (rasom) [inline src/status_im/data_store/messages.cljs:71]: Why these changes are necessary for PR?

File: src/status_im/data_store/messages.cljs [hunk 24]
```diff
@@ -84,52 +97,48 @@
...
-                                              :received-timestamp (datetime/timestamp)})))))
-(defn delete
+(defn save-message-tx
+  "Returns tx function for saving message"
+  [{:keys [message-id from] :as message}]
+  (fn [realm]
+    (when-not (core/exists? realm :message :message-id message-id)
+      (core/create realm
+                   :message
+                   (prepare-message (merge default-values message {:from (or from "anonymous")}))))))
```
Comment (rasom) [inline src/status_im/data_store/messages.cljs:107]: 80 chars limit

File: src/status_im/data_store/messages.cljs [hunk 25]
```diff
@@ -84,52 +97,48 @@
...
   [message-id]
-  (when (data-store/exists? message-id)
-    (data-store/delete message-id)))
-
-(re-frame/reg-fx
-  :data-store/save-message
-  (fn [message]
-    (async/go (async/>! core/realm-queue #(save message)))))
+  (fn [realm]
+    (when-let [message (core/single (core/get-by-field realm :message :message-id message-id))]
```
Comment (rasom) [inline src/status_im/data_store/messages.cljs:113]: 80 chars limit

File: src/status_im/data_store/messages.cljs [hunk 26]
```diff
@@ -84,52 +97,48 @@
...
-    (async/go (async/>! core/realm-queue #(delete message-id)))))
-
-(defn update-message
+(defn update-message-tx
+  "Returns tx function for updating message"
   [{:keys [message-id] :as message}]
-  (when-let [{:keys [chat-id]} (data-store/get-by-id message-id)]
-    (data-store/save (prepare-message (assoc message :chat-id chat-id)))))
+  (fn [realm]
+    (when-let [{:keys [chat-id] :as loaded} (some-> (core/get-by-field realm :message :message-id message-id)
```
Comment (rasom) [inline src/status_im/data_store/messages.cljs:120]: 80 chars limit

File: src/status_im/chat/events.cljs (hunk 11 above)
Comment (cammellos) [inline src/status_im/chat/events.cljs:115]: removed semicolon

File: src/status_im/data_store/requests.cljs [hunk 27]
```diff
@@ -1,20 +1,25 @@
...
+  [request]
+  (fn [realm]
+    (core/create realm :request request true)))
 
-(re-frame/reg-fx
-  :data-store/mark-request-as-answered
-  (fn [{:keys [chat-id message-id]}]
-    (async/go (async/>! core/realm-queue #(data-store/mark-as-answered chat-id message-id)))))
+(defn mark-request-as-answered-tx
+  "Given chat-id and message-id, returns tx function for marking request as answered"
```
Comment (rasom) [inline src/status_im/data_store/requests.cljs:19]: 80 chars limit

File: src/status_im/data_store/transport.cljs [hunk 28]
```diff
@@ -22,23 +20,27 @@
...
-
-(re-frame/reg-fx
-  :data-store.transport/delete
-  (fn [chat-id]
-    (async/go (async/>! core/realm-queue #(data-store/delete chat-id)))))
+(defn delete-transport-tx
+  "Returns tx function for deleting transport"
+  [chat-id]
+  (fn [realm]
+    (let [transport (core/single (core/get-by-field realm :transport :chat-id chat-id))]
```
Comment (rasom) [inline src/status_im/data_store/transport.cljs:45]: 80 chars limit

File: src/status_im/transport/core.cljs [hunk 29]
```diff
@@ -27,13 +28,13 @@
...
           topic (transport.utils/get-topic constants/contact-discovery)]
       (handlers-macro/merge-fx cofx
-                         {:shh/add-discovery-filter {:web3           web3
-                                                     :private-key-id public-key
-                                                     :topic topic}
-                          :shh/restore-sym-keys {:web3       web3
-                                                 :transport  (:transport/chats db)
-                                                 :on-success sym-key-added-callback}}
-                         (inbox/initialize-offline-inbox)))))
+                               {:shh/add-discovery-filter {:web3           web3
```
Comment (rasom) [inline src/status_im/transport/core.cljs:31]: unnecessary changes

File: src/status_im/transport/message/v1/group_chat.cljs [hunk 30]
```diff
@@ -31,10 +31,10 @@
                                               :sym-key    sym-key
                                               :on-success (fn [sym-key sym-key-id]
                                                             (re-frame/dispatch [:group/add-new-sym-key {:chat-id    chat-id
-                                                                                                   :signature  signature
-                                                                                                   :sym-key    sym-key
-                                                                                                   :sym-key-id sym-key-id
-                                                                                                   :message    message}]))}}
+                                                                                                        :signature  signature
```
Comment (rasom) [inline src/status_im/transport/message/v1/group_chat.cljs:34]: unnecessary changes

File: src/status_im/transport/message/v1/group_chat.cljs [hunk 31]
```diff
@@ -46,9 +46,9 @@
     {:shh/get-new-sym-key {:web3 (get-in cofx [:db :web3])
                            :on-success (fn [sym-key sym-key-id]
                                          (re-frame/dispatch [:group/send-new-sym-key {:chat-id    chat-id
-                                                                                 :sym-key    sym-key
-                                                                                 :sym-key-id sym-key-id
-                                                                                 :message    message}]))}}))
+                                                                                      :sym-key    sym-key
```
Comment (rasom) [inline src/status_im/transport/message/v1/group_chat.cljs:49]: unnecessary changes

File: src/status_im/transport/message/v1/group_chat.cljs [hunk 32]
```diff
@@ -93,8 +93,8 @@
             (if (removed me) ;; we were removed
               (handlers-macro/merge-fx cofx
                                  (models.message/receive
-                                   (models.message/system-message chat-id random-id now
```
Comment (rasom) [inline src/status_im/transport/message/v1/group_chat.cljs:96]: unnecessary changes

File: src/status_im/transport/message/v1/public_chat.cljs [hunk 33]
```diff
@@ -25,18 +26,17 @@
...
-       :shh/add-filter            {:web3       web3
-                                   :sym-key-id sym-key-id
-                                   :topic      (transport.utils/get-topic chat-id)
-                                   :chat-id    chat-id}
-       :data-store.transport/save {:chat-id chat-id
-                                   :chat    (-> (get-in db [:transport/chats chat-id])
-                                                (assoc :sym-key-id sym-key-id)
-                                                ;;TODO (yenda) remove once go implements persistence
-                                                (assoc :sym-key sym-key))}}))))
+  ::add-new-sym-key
```
Comment (rasom) [inline src/status_im/transport/message/v1/public_chat.cljs:29]: unnecessary changes

**Review by cammellos — APPROVED** [review]
looks sound to me.
 I kind of liked the separate namespaces for the query (or at least the separation between queries), but that's ok. 
very good effort.

**Review by rasom — CHANGES_REQUESTED** [review]

File: src/status_im/chat/events.cljs (hunk 11 above)
Comment (cammellos) [reply src/status_im/chat/events.cljs:115]: not semicolon, double quote :)

File: src/status_im/ui/screens/group/chat_settings/events.cljs (hunk 7 above)
Comment (janherich) [reply src/status_im/ui/screens/group/chat_settings/events.cljs:19]: So I should **re-introduce** all those trailing whitespaces, correct ?

**Comment by rasom** [comment]
So just to make it clear, the best way to introduce such a big changeset (~ +700 -1000) (**though only in my opinion**) is to split it into readable meaningful chunks. In this particular case we could have at least four separate PRs (or at least commits), where:
1) the first one which actually introduces the new functionality (if it's separate PR also shows how this new functionality is applied in some particular place in app, in this case examples for `:data-store/tx` and `:data-store/base-tx` effects)
2) another one which applies these effects to the rest parts of code
3) another one which contains any refactoring related to fetching data or something
4) another one which contains fixes to codestyle if necessary 

In this case it's clear what is important in PR, what should be read and not missed accidentaling during scrolling.

File: src/status_im/ui/screens/group/chat_settings/events.cljs (hunk 7 above)
Comment (cammellos) [reply src/status_im/ui/screens/group/chat_settings/events.cljs:19]: @rasom I think at this point it's best to review this PR as it is, reviewing a moving target is almost more difficult than reviewing it with those changes. 
What do you think? 

To avoid this in the future we can automatically lint or prevent pushing unlinted code in a githook, @janherich seems to have the best editor setup and maybe can help. A PR that lints all the files will be necessary. 

I think the point is well taken about unnecessary changes :) We can have further discussion about this though

**Comment by janherich** [comment]
@rasom What I don't get is why we are so super strict about unnecessary changes, even if it's just couple of lines which correct wrong indentation (automatically done by good editor) and not at all so strict when such changes (code not adhering to clojure styleguide) are introduced in the first place ?

**Comment by yenda** [comment]
@janherich you accepted the PR so you are the one with the answer :rofl:  https://github.com/status-im/status-react/pull/3962

**Comment by janherich** [comment]
@yenda Yes, my answer would be not to be super strict just for the sake of it, just as I was not while reviewing #3962 :) 
I just didn't know that once incorrect indentation lands in codebase, it's so hard to fix it later, next time, I will be super strict while reviewing as well :)

**Comment by janherich** [comment]
@rasom Can you please state clearly in the advance which parts of PR should be excluded (so I can open later PR with those changes) so it's fine for you to approve it ? I already spend quite some time in the morning "fixing" the ident things only to be rejected again, so I would like to know what exactly is problematic, preferably all in advance, so I don't have to revisit it again and again.

**Comment by mandrigin** [comment]
@janherich I'll test this branch's performance implications tomorrow. Will report here.

**Comment by janherich** [comment]
Great, thanks @mandrigin !

**Comment by rasom** [comment]
@janherich i suppose this comment explains what i would be happy to see https://github.com/status-im/status-react/pull/4079#issuecomment-385908916 

> This PR address the point by adding :data-source/tx and :data-source/base-tx effects (for account/base realm respectively) + reworking the data-source.ENTITY namespaces to expose public functions modelling transaction operations.

so we definitely could have one PR with two commits where one will introduce ":data-source/tx and :data-source/base-tx" and another one applies them where necessary

> I also removed all the data-source.realm.ENTITY namespace, as that was just unnecessary burden, which resulted in quite a big code size reduction (almost 300LOC less).

this is refactoring, can live in a separate PR

all indentation changes, at least in forms (i mean clojure forms) which are not touched by previous PRs can be introduced by another PR

And that's it, does it make sense?

**Comment by janherich** [comment]
@rasom It will be hard to retroactively re-introduce `data-source.realm.ENTITY` namespaces such that number of changes (eq due to formatting) will minimal, but I will try to do my best. The PR is already squashed, so I can't just select some of my previous commits.

**Comment by mandrigin** [comment]
@janherich @rasom I wasn't able to properly compare the performance of this PR and the current develop.
The reason being visible degradation of performance when you receive more messages.

See the table.


*msgs* column shows how many messages are in the cache at this moment.

When the cache is empty, the code processes 50 messages received essentially at once in ~6 seconds.
When there are ~500 already received messages in a chatroom, it degrades to ~20 seconds for the same operation.

I used `./gradlew assebleRelease` to make a measing build + added `debuggable:true` to `AndroidManifest.xml`.
Galasxy S6, Android 7.0

**Comment by jeluard** [comment]
@janherich What is the status on this PR?

**Comment by janherich** [comment]
@jeluard I will close it, as half of the work is already accomplished by another recently merged PR and rest will be easier to do through follow-up PR.
